File: dspark/generator.py
"""
Fast Generator Engine for DSpark (Gemini integration).
"""

from typing import Optional
from .client import GeminiClient


DRAFT_SYSTEM_INSTRUCTION = """You are a high-speed software development assistant.
Generate clean, idiomatic, and complete code implementations according to the requested prompt.
Do not omit details or leave placeholders (like '// TODO' or 'pass') unless explicitly requested.
Wrap your output in standard markdown code blocks.
"""


class GeminiGenerator:
    """
    High-throughput draft code generator powered by Google Gemini.
    """

    def __init__(self, client: Optional[GeminiClient] = None, model: Optional[str] = None):
        self.client = client or GeminiClient()
        self.model = model or self.client.default_model

    def generate_draft(
        self,
        prompt: str,
        language: Optional[str] = None,
        temperature: float = 0.7,
    ) -> str:
        """
        Generate a fast, high-throughput code draft from a natural language specification.
        """
        lang_directive = f" (Target language: {language})" if language else ""
        full_prompt = f"Implement the following feature or function{lang_directive}:\n\n{prompt}"

        return self.client.generate_content(
            prompt=full_prompt,
            system_instruction=DRAFT_SYSTEM_INSTRUCTION,
            model=self.model,
            temperature=temperature,
        )

File: src/generator.rs
//! High-throughput draft generator (Gemini / any ModelClient).

use crate::client::{ClientError, ModelClient};
use crate::prompts::DRAFT_SYSTEM_INSTRUCTION;
use crate::util::extract_code_blocks;

pub struct GeminiGenerator {
    client: ModelClient,
}

impl GeminiGenerator {
    pub fn new() -> Result<Self, ClientError> {
        Ok(Self {
            client: ModelClient::from_spec(
                &std::env::var("GEMINI_MODEL").unwrap_or_else(|_| "gemini-2.5-flash".into()),
            )?,
        })
    }

    pub fn with_spec(spec: &str) -> Result<Self, ClientError> {
        Ok(Self {
            client: ModelClient::from_spec(spec)?,
        })
    }

    pub fn with_client(client: ModelClient) -> Self {
        Self { client }
    }

    pub async fn generate_draft(
        &self,
        prompt: &str,
        language: Option<&str>,
        temperature: f32,
    ) -> Result<String, ClientError> {
        let lang_directive = language
            .map(|l| format!(" (Target language: {})", l))
            .unwrap_or_default();
        let full_prompt = format!(
            "Implement the following feature or function{lang_directive}:\n\n{prompt}"
        );
        let raw = self
            .client
            .complete(
                &full_prompt,
                Some(DRAFT_SYSTEM_INSTRUCTION),
                temperature,
                false,
            )
            .await?;
        Ok(extract_code_blocks(&raw))
    }
}
